app/services/credentials_verifier.go

```diff
@@ -45,5 +47,16 @@ func CredentialsVerifier(store data.AccountStore, cfg *app.Config, username stri
 		return nil, FieldErrors{{"credentials", ErrExpired}}
 	}
 
+	//Check TOTP MFA
+	if account.TOTPEnabled() {
+		secret, err := compat.Decrypt([]byte(account.TOTPSecret.String), cfg.DBEncryptionKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "TOTPDecrypt")
+		}
+		if !totp.Validate(totpCode, secret) {
+			return nil, FieldErrors{{"totp", ErrInvalidOrExpired}}
```

Comment:
Should we vary the error here if MFA code is missing?  Thinking mostly in terms of how applications might try submitting without MFA then prompt user for a code based on the response.  This also comes into play if we want to add HOTP support that requires out-of-band code delivery.

Note the timing above we should keep going at least through password validation when making any changes.

Reply:
I think probably not - maybe a back end api could support a status check on a user though.

Reply:
Hmm yes a constant time comparison that doesn't reveal whether the password or OTP was correct might be ideal. However if there's any chance something is revealed, I'd prefer to reveal the correctness of the OTP.

Reply:
It is already returning invalid or expired for the OTP.  Added missing in 153d385a2fee06c0256a6a3863d9d92b71e2b2b8